mosaic-runtime/src/commonMain/kotlin/com/jakewharton/mosaic/mosaic.kt

```diff
@@ -53,86 +54,112 @@ public fun renderMosaic(content: @Composable () -> Unit): String {
 }
 
 public suspend fun runMosaic(content: @Composable () -> Unit) {
-	runMosaic(content, ::platformDisplay)
+	coroutineScope {
+		val mosaicComposition = MosaicComposition(coroutineScope = this, onRender = ::platformDisplay)
+		mosaicComposition.setContent(content)
+		mosaicComposition.awaitComplete()
+	}
 }
 
-internal suspend fun runMosaic(
-	content: @Composable () -> Unit,
-	display: (CharSequence) -> Unit,
-): Unit = coroutineScope {
-	val terminal = MordantTerminal()
+internal class MosaicComposition(
+	coroutineScope: CoroutineScope,
+	override val onRender: (CharSequence) -> Unit,
+) : BaseMosaicComposition(coroutineScope) {
+	private val terminal = MordantTerminal()
 
-	val rendering = if (debugOutput) {
+	override val rendering: Rendering = if (debugOutput) {
 		@OptIn(ExperimentalTime::class) // Not used in production.
 		DebugRendering(ansiLevel = terminal.info.ansiLevel.toMosaicAnsiLevel())
 	} else {
 		AnsiRendering(ansiLevel = terminal.info.ansiLevel.toMosaicAnsiLevel())
 	}
 
-	val clock = BroadcastFrameClock()
-	val job = Job(coroutineContext[Job])
-	val composeContext = coroutineContext + clock + job
+	override val terminalInfo: MutableState<Terminal> =
+		mutableStateOf(Terminal(size = IntSize(terminal.info.width, terminal.info.height)))
 
-	GlobalSnapshotManager.ensureStarted(composeContext)
+	init {
+		coroutineScope.sendFrames()
+		coroutineScope.updateTerminalInfo()
+	}
 
-	launch(composeContext) {
-		while (true) {
-			clock.sendFrame(0L) // Frame time value is not used by Compose runtime.
-			delay(50L)
+	private fun CoroutineScope.updateTerminalInfo() {
+		launch(composeContext) {
+			while (true) {
+				val currentTerminalInfo = terminalInfo.value
+				if (terminal.info.updateTerminalSize() &&
+					(
+						currentTerminalInfo.size.width != terminal.info.width ||
+							currentTerminalInfo.size.height != terminal.info.height
+						)
+				) {
+					terminalInfo.value = Terminal(size = IntSize(terminal.info.width, terminal.info.height))
+				}
+				delay(50L)
+			}
 		}
 	}
+}
 
-	val recomposer = Recomposer(composeContext)
-	launch(composeContext, start = CoroutineStart.UNDISPATCHED) {
-		recomposer.runRecomposeAndApplyChanges()
+internal abstract class BaseMosaicComposition(coroutineScope: CoroutineScope) {
```

Comment:
Would prefer this be called `MosaicComposition` and the subtypes be named with something else like `TestMosaicComposition` and `RenderingMosaicComposition` (or something)